piper/buttondialog.py

```diff
@@ -102,13 +128,112 @@ def _init_mapping_page(self, buttons):
+                                         False, None, None, None, None)
+        if status != Gdk.GrabStatus.SUCCESS:
+            return False
+        self._grab_pointer = pointer
+        self.grab_add()
+        return True
+
+    def _release_grab(self):
+        # Releases a previously grabbed keyboard seat, if any.
+        if self._grab_pointer is None:
+            return
+        self._grab_pointer.get_seat().ungrab()
+        self._grab_pointer = None
+        self.grab_remove()
+
+    def _set_keystroke(self):
+        # Sets the current action type to "key" and caches the keystroke in a
+        # ratbagd compatible form, then restores the overview page and releases
+        # the grab.
+        self._action_type = "key"
+        self._key_mapping = self._keystroke.to_evdev()
+        self.stack_mapping.set_visible_child_name("overview")
+        self._release_grab()
+
+    def do_key_press_event(self, event):
+        # Overrides Gtk.Widget's standard key press event callback, so we can
+        # capture the pressed buttons in capture mode. Gratefully copied from
+        # GNOME Control Center's keyboard panel.
+        # Don't process key events when we're not in capture mode.
+        if self.stack_mapping.get_visible_child_name() == "overview":
+            return Gtk.Widget.do_key_press_event(self, event)
+
+        # TODO: remove this workaround when libratbag removes its keycode
+        # contraints. When that happens, we just cache all keypresses in the
+        # order they arrive and set the keystroke upon Return.
+        # GdkEventKey.is_modified isn't exposed through PyGObject (see
+        # https://bugzilla.gnome.org/show_bug.cgi?id=752784), so we have to
+        # approximate its behaviour ourselves. This selection is from Gtk's
+        # default mod mask and should be fine for now for most use cases.
+        event.is_modifier = event.keyval in self._MODIFIERS
+        keyval_lower = Gdk.keyval_to_lower(event.keyval)
+
+        # We only want to bind keystrokes using the default modifiers.
+        real_mask = event.state & Gtk.accelerator_get_default_mod_mask()
+
+        # Normalize tab.
+        if keyval_lower == Gdk.KEY_ISO_Left_Tab:
+            keyval_lower = Gdk.KEY_Tab
+
+        # Put shift back if it changed the case of the key, not otherwise.
+        if keyval_lower != event.keyval:
+            real_mask |= Gdk.ModifierType.SHIFT_MASK
+
+        # HACK: we don't want to use SysRq as a keybinding, but we do want
+        # Al+Print, so we avoid translating Alt+Print to SysRq.
+        if keyval_lower == Gdk.KEY_Sys_Req and (real_mask & Gdk.ModifierType.MOD1_MASK):
+            keyval_lower = Gdk.KEY_Print
+
+        # Backspace disables the current shortcut.
+        if not event.is_modifier and real_mask == 0 and keyval_lower == Gdk.KEY_BackSpace:
+            self._keystroke.clear()
+            self._set_keystroke()
+            return Gdk.EVENT_STOP
+
+        if event.is_modifier:
```

Comment:
As said above, this if/else can go once libratbag doesn't care about its key mapping signature anymore.